adala/utils/parse.py

```diff
@@ -21,102 +20,78 @@
-                    auto_arg_index=auto_arg_index,
-                )
-
-                # format the object and append to the result
-                # HACK: if the format_spec is invalid, assume this field_name was not meant to be a variable, and don't substitute anything
-                formatted_field = self.format_field(obj, format_spec)
-                if formatted_field is None:
-                    result.append("{" + ":".join([field_name, format_spec]) + "}")
-                else:
-                    result.append(formatted_field)
+    def format_field(self, value, format_spec):
+        if isinstance(value, str) and value.startswith("{"):
+            # This is a preserved placeholder, return as is
+            if format_spec:
+                return value[:-1] + ":" + format_spec + "}"
+            else:
+                return value
 
-        return "".join(result), auto_arg_index
+        try:
+            return super().format_field(value, format_spec)
+        except (ValueError, TypeError):
+            # If format spec is invalid, preserve the original field name and format spec
+            if format_spec:
+                return "{" + self._current_field_name + ":" + format_spec + "}"
+            return str(value)
 
 
 PartialStringFormat = PartialStringFormatter()
 
 
-def partial_str_format(string, **kwargs):
+def partial_str_format(format_string: str, **kwargs) -> str:
     """
-    Formats a string with a subset of the arguments.
-    Analogous to str.format, but ignores missing arguments.
+    Format a string with the provided variables while preserving any unprovided placeholders.
+    Preserves format specifiers for both provided and unprovided variables.
+
+    Args:
+        format_string: The string to format
+        **kwargs: The variables to use for formatting
+
+    Returns:
+        The formatted string with preserved unprovided placeholders and format specifiers
+
+    Examples:
+        >>> partial_str_format("Hello {name}!", name="World")
+        'Hello World!'
+        >>> partial_str_format("Hello {name} {unknown}!", name="World")
+        'Hello World {unknown}!'
+        >>> partial_str_format("Value: {x:.2f}", x="not_a_float")
+        'Value: {x:.2f}'
     """
-    return PartialStringFormat.format(string, **kwargs)
+    if not format_string:
+        return ""
+
+    # Temporarily replace valid format strings to protect them from escaping
+    format_pattern = re.compile(r"\{[^{}]+\}")
+    markers = {
+        f"__MARKER_{i}__": m.group(0)
+        for i, m in enumerate(format_pattern.finditer(format_string))
+    }
+
+    processed = format_string
+    for marker, format_str in markers.items():
+        processed = processed.replace(format_str, marker)
+
+    # Escape remaining brackets and restore format strings
+    processed = processed.replace("{", "{{").replace("}", "}}")
+    for marker, format_str in markers.items():
+        processed = processed.replace(marker, format_str)
```

Comment:
I hate to say it, but could we test that the performance isn't heavily impacted by this? I'm somewhat worried about all of this string formatting
At the very least, can we at least smoke test that this isn't noticeably longer for a few projects?

Reply:
Ideally, including some pathological cases, like Max's project and/or some enhanced prompts

Reply:
Good idea, will test on Max's project

Reply:
On sample 20 tasks in Max's project, I get 3:35 on this branch vs 3:45 on master with the same outputs (errors on the same 2 tasks), so within expected margin of error for no difference :+1: 
Good callout though